### ISSUE TITLE
Auto-Recognize CSV Separator Character

### ISSUE THREAD

**Issue by tm-ms** [microstream-one/microstream#59]
Confusion with the legacy type mapping example showed that whenever CSV files are used, the separator has to be recognized automatically. Otherwise, a lot of people will have frustratingly big difficulties using a CSV file with MicroStream.

Background:
By default, MicroStream uses tabs as separator characters.
The tab character was created for exactely this usage and it is the best (and only reasonable) character for the job. It allows to create a logical table structure in a text file, with a dynamically adjustable column width. All just in a simple ASCII value. This makes table-like text files (that CSV is) indefinitely more human-readable while having no drawback at all (still one character, just like comma or semicolon is).
Since CSV initially stood for "comma-separated values", but can also be read as "character-separated values" (and since all kinds of characters are used as separators, anyway), this does no really violate the nature of CSV.
In any case, it is downright idiotic to use a comma or a semicolon (or any other character) instead.
Hence the tab.

HOWEVER

Displaying the content of a tab-separated CSV file on a website very often replaces the tabs by spaces. Either automatically or indirectly forced to be done manually.
This is also the case in our documentation.
See "refactorings.csv" here: https://manual.docs.microstream.one/data-store/legacy-type-mapping

That way, the reader has no (easy) way to recognize that the file is meant to be separated by tabs.
And even if there is such a way (the documentation can include a hint about the matter, see https://github.com/microstream-one/docs-manual/issues/5), there will still be a lot of people who can't differentiate between a tab and a space or use multiple tabs instead of properly adjusting the dynamic tab with in their editor, etc. A lot of people will probably defiantly use comma or semicolon and will blame MicroStream if it doesn't accept the idiotic separator.

To avoid tons of confusion, frustration, support work, etc. in the future, MicroStream must get "smart enough" here to cover all cases automatically.

An pragmatic automatic recognition should be rather easy to be implemented:
Scan all code points (~chars) of the input string from line to line and count the occurances of ";", "," and "\t". If one consistently appears with a certain minimum amount in every line, this is most probably the separator character. Even strings containing false positives shouldn't falsify the total result too much.

Also, the exception message can be improved to give a hint regarding the accepted seperator values.

Additionally, a CsvConfiguration (already exists) can be optionally passed.
Also a simplified version with just the separator character.
And if the file ends with ".tsv" (see https://en.wikipedia.org/wiki/Tab-separated_values), the separator to be used is clear, anyway.

This might sound like a lot of work, but with the already existing CSV parsing framework in the MicroStream Base code, this is just a few lines of code and the string-scanning loop (actually foreach method call).

**Comment by tm-ms** [comment]
The XCSV framework now magically guesses the value separator of a csv string if no concrete value is specified.

There are now the two methods XCSV#parse and XCSV#readFromFile with the following 4 variants each:
1.) String/File only
This tries to read the metacharacter definitions from the first line of the data if present. Otherwise, it does the magic trick of guessing the separator and using XCsvDataType.XCSV.configuration() for everything else.

2.) String/File, XCsvDataType
Like 1), but uses the preferences of the passed dataType. E.g. CSV with a bias to "," instead of "\t".

3.) String/File, char valueSeparator
Uses the specified valueSeparator without guessing and uses XCsvDataType.XCSV.configuration() for everything else.

4.) String/File, XCsvConfiguration
Like 3), but uses the passed configuration directly

The Main class for testing the guessing is `MainTestXCsvSeparatorGuessing`.
4 tiny examples of test data are contained there as a comment.

Any additional test is welcome, especially ones that cause the guessing to fail.
It won't always be correct since it uses heuristics instead of strictly defined rules, but it should get all cases within reason of a real-world application right.

**Comment by zdenek-jonas** [comment]
Implementation of this issue causes a side effect when using cvs legacy type mapping.

Perhaps the dots in this file are incorrectly identified as a separator

csv file:
```
old	current
test.microstream.legacy.csv.data.CsvPerson	test.microstream.legacy.csv.data.CsvPerson2
test.microstream.legacy.csv.data.CsvPerson#original	test.microstream.legacy.csv.data.CsvPerson2#copy
```

call stack: 
```
java.lang.IllegalArgumentException: Invalid column count in row 1 (1 required, 3 available)

	at one.microstream.chars.StringTable$Default.validateColumnCount(StringTable.java:82)
	at one.microstream.chars.StringTable$Default.<init>(StringTable.java:125)
```

**Comment by tm-ms** [comment]
Interesting:
I thought at first the many peroids compared to the low amount of tabs would throw off the heuristics algorithm.
But that is not the case. It still weighs the tabs nicely higher than the peroids (because the peroids have a line without occurances which is a major malus in the formula).
Instead, the "#" is weighed "Infinity".
```
";" = 0.0 (weight = 1.2)
"," = 0.0 (weight = 1.3)
"|" = 0.0 (weight = 1.0)
":" = 0.0 (weight = 0.9)
"~" = 0.0 (weight = 0.9)
"*" = 0.0 (weight = 0.8)
"-" = 0.0 (weight = 0.8)
"." = 0.923077 (weight = 0.8)
"\t" = 1.1 (weight = 1.1)
"#" = Infinity (weight = 0.9)
```

This is both hilarious and completely wrong. Must be some funny special case. I'll fix it.
...
But I'll throw out the "." nevertheless. Especially for Java full qualified types like in refactoring mappings, the risk of too many periods outweighing the actual separator is too high.
Whoever absolutely has to use "." as a value separator must take the enormous effort to write an additional `, '.'` when importing tabular data...

**Comment by tm-ms** [comment]
The reason was that minCount and maxCount were not updated for the last line.
If a value separator candidate only occured in the last line, its total count was > 0, but its minCount was still the initial value of max_int. In the formula (* 1/minCount), this resulted in a evaluation score of "infinite".
I fixed it by rearranging the loop condition so that the statistics get update for both the new line case and the end of data case.

This eliminated the problem.

I also reduced the valid value separators to a reasonable minimum to prevent false positives as good as possible.
The list now contains only the 5 characters `";", ",", "\t", "|", ":"`
